File: memory-jar-backend/src/main/java/com/memoryjar/memoryjarbackend/model/Memory.java
package com.memoryjar.memoryjarbackend.model;

import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import jakarta.validation.constraints.FutureOrPresent;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.LinkedHashSet;
import java.util.Set;

@Entity
@Table(
        name = "memories",
        indexes = @Index(
                name = "idx_memories_status_delivery_date",
                columnList = "status, delivery_date"
        )
)
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Memory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(nullable = false)
    private String title;

    @NotBlank
    @jakarta.persistence.Convert(converter = com.memoryjar.memoryjarbackend.util.StringEncryptorConverter.class)
    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MemoryType type;

    @Column(name = "media_url")
    private String mediaUrl;

    @NotNull
    @FutureOrPresent
    @Column(name = "delivery_date", nullable = false)
    private LocalDateTime deliveryDate;

    @ElementCollection(targetClass = DeliveryChannel.class)
    @CollectionTable(name = "memory_delivery_channels", joinColumns = @JoinColumn(name = "memory_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "channel", nullable = false)
    @NotEmpty
    @Builder.Default
    private Set<DeliveryChannel> deliveryChannels = new LinkedHashSet<>();

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private MemoryStatus status = MemoryStatus.PENDING;

    @NotBlank
    @Column(name = "recipient_identifier", nullable = false)
    private String recipientIdentifier;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "delivered_at")
    private LocalDateTime deliveredAt;

    @Column(name = "last_error", columnDefinition = "TEXT")
    private String lastError;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
        if (status == null) {
            status = MemoryStatus.PENDING;
        }
    }

    public enum MemoryType {
        TEXT, AUDIO, IMAGE, VIDEO, DOCUMENT
    }

    public enum DeliveryChannel {
        WHATSAPP, PHONE_CALL, SMS, EMAIL, PUSH_NOTIFICATION
    }

    public enum MemoryStatus {
        PENDING, DELIVERED, FAILED
    }
}

File: memory_jar_mobile/lib/models/memory.dart
enum MemoryType { TEXT, AUDIO, IMAGE, VIDEO, DOCUMENT }
enum DeliveryChannel { WHATSAPP, PHONE_CALL, SMS, EMAIL, PUSH_NOTIFICATION }
enum MemoryStatus { PENDING, DELIVERED, FAILED }

class Memory {
  final int? id;
  final String title;
  final String content;
  final MemoryType type;
  final String? mediaUrl;
  final DateTime deliveryDate;
  final List<DeliveryChannel> deliveryChannels;
  final MemoryStatus status;
  final String recipientIdentifier;
  final DateTime? createdAt;
  final DateTime? deliveredAt;
  final String? lastError;

  Memory({
    this.id,
    required this.title,
    required this.content,
    required this.type,
    this.mediaUrl,
    required this.deliveryDate,
    required this.deliveryChannels,
    this.status = MemoryStatus.PENDING,
    required this.recipientIdentifier,
    this.createdAt,
    this.deliveredAt,
    this.lastError,
  });

  factory Memory.fromJson(Map<String, dynamic> json) {
    return Memory(
      id: json['id'],
      title: json['title'],
      content: json['content'],
      type: MemoryType.values.byName(json['type']),
      mediaUrl: json['mediaUrl'],
      deliveryDate: DateTime.parse(json['deliveryDate']),
      deliveryChannels: (json['deliveryChannels'] as List)
          .map((e) => DeliveryChannel.values.byName(e))
          .toList(),
      status: MemoryStatus.values.byName(json['status']),
      recipientIdentifier: json['recipientIdentifier'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      deliveredAt: json['deliveredAt'] != null ? DateTime.parse(json['deliveredAt']) : null,
      lastError: json['lastError'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'type': type.name,
      'mediaUrl': mediaUrl,
      'deliveryDate': deliveryDate.toIso8601String(),
      'deliveryChannels': deliveryChannels.map((e) => e.name).toList(),
      'status': status.name,
      'recipientIdentifier': recipientIdentifier,
      'createdAt': createdAt?.toIso8601String(),
      'deliveredAt': deliveredAt?.toIso8601String(),
      'lastError': lastError,
    };
  }
}

extension MemoryTypeLabel on MemoryType {
  String get label {
    return switch (this) {
      MemoryType.TEXT => 'Text',
      MemoryType.AUDIO => 'Voice',
      MemoryType.IMAGE => 'Photo',
      MemoryType.VIDEO => 'Video',
      MemoryType.DOCUMENT => 'Document',
    };
  }
}

extension DeliveryChannelLabel on DeliveryChannel {
  String get label {
    return switch (this) {
      DeliveryChannel.WHATSAPP => 'WhatsApp',
      DeliveryChannel.PHONE_CALL => 'Phone call',
      DeliveryChannel.SMS => 'SMS',
      DeliveryChannel.EMAIL => 'Email',
      DeliveryChannel.PUSH_NOTIFICATION => 'Push',
    };
  }
}

extension MemoryStatusLabel on MemoryStatus {
  String get label {
    return switch (this) {
      MemoryStatus.PENDING => 'Sealed',
      MemoryStatus.DELIVERED => 'Delivered',
      MemoryStatus.FAILED => 'Needs attention',
    };
  }
}
